nethack/src/board.c

```diff
@@ -0,0 +1,253 @@
+
+#include "nhcurses.h"
+#include <time.h>
+
+static struct gamewin *gw;
+static struct win_menu *mdat;
+static nh_bool done;
+static struct nh_board_entry *entries;
+static int listlen, maxplayersize;
+
+static void
+board_display_curses_ui(struct nh_menuitem *items, int icount)
+{
+    int startx, starty, x1, x2, y1, y2;
+
+    if (gw != NULL) {
+        delwin(mdat->content);
+        delwin(gw->win);
+        delete_gamewin(gw);
+    }
+
+    gw = alloc_gamewin(sizeof (struct win_menu));
+    gw->draw = draw_menu;
+    gw->resize = resize_menu;
+    mdat = (struct win_menu *)gw->extra;
+    mdat->items = items;
+    mdat->icount = icount;
+    mdat->title = "Binge Board";
+    mdat->selected = calloc(icount, sizeof (nh_bool));
+    if (!mdat->selected)
+        curses_msgwin("Couldn't allocate memory for the binge board");
+        return;
+    mdat->x1 = 0;
+    mdat->y1 = 0;
+    mdat->x2 = -1;
+    mdat->y2 = -1;
+
+    assign_menu_accelerators(mdat);
+    layout_menu(gw);
+
+    startx = (COLS - mdat->width) / 2;
+    starty = (LINES - mdat->height) / 2;
+
+    gw->win = newwin(mdat->height, mdat->width, starty, startx);
+    if (gw->win == NULL) {
+        curses_msgwin("Error: newwin returned NULL!");
+        return;
+    }
+    if (keypad(gw->win, TRUE) == ERR) {
+        curses_msgwin("Error: keypad for gw->win errored out!");
+        return;
+    }
+    if (wattron(gw->win, FRAME_ATTRS) == ERR) {
+        curses_msgwin("Error: wattron for gw->win errored out!");
+        return;
+    }
+    box(gw->win, 0, 0);
+    if (wattroff(gw->win, FRAME_ATTRS) == ERR) {
+        curses_msgwin("Error: wattroff for gw->win errored out!");
+        return;
+    }
+    mdat->content =
+        derwin(gw->win, mdat->innerheight, mdat->innerwidth,
+               mdat->frameheight - 1, 2);
+    if (mdat->content == NULL) {
+        curses_msgwin("Error: derwin for gw->win errored out!");
+        return;
+    }
+    if (leaveok(gw->win, TRUE) == ERR) {
+        curses_msgwin("Error: leaveok for gw->win errored out!");
+        return;
+    }
+    if (leaveok(mdat->content, TRUE) == ERR) {
+        curses_msgwin("Error: leaveok for mdat->content errored out!");
+        return;
+    }
+    wtimeout(gw->win, 500);
+}
+
+static void
+board_add_entry(struct nh_board_entry *entry, struct nh_menuitem **items,
+                int *size, int *icount, int maxplayerwidth,
+                enum nh_menuitem_role txtrole)
+{
+    char fmtline[BUFSZ], line[BUFSZ], name[BUFSZ], hptxt[BUFSZ], entxt[BUFSZ];
+    struct tm tm;
+    time_t t, now;
+    char *inactive;
+
+    now = time(NULL);
+
+    strptime(entry->lastactive, "%Y-%m-%d %H:%M:%S.000000%z", &tm);
+    t = mktime(&tm);
+
+    if(difftime(now, t) > 30) { //Marks a player inactive after 30 seconds
+        inactive = "I";
+    } else {
+        inactive = "";
+    }
+
+    snprintf(name, 60, "%s the %s %s %s %s", entry->name, entry->plalign,
```

Comment:
Instead of using calculated magic numbers as the length here, just pass in the actual max size of the buffer `BUFSZ`